### PR TITLE
ESDC1A Documentation + Implementation

### PR DESCRIPTION
**Pull request by lukelowry**
## Description

Implementation and documentation for the `ESDC1A` PhasorDynamics exciter model.

## Proposed changes

- Added `ESDC1A` model implementation, documentation, and diagram.
- Added `ESDC1A` JSON parsing and `SystemModel` assembly support.
- Added focused `PhasorDynamicsExciterEsdc1aTest` unit coverage.
- Added sparse Jacobian support through Enzyme and DependencyTracking build paths.
- Updated exciter registration, input-format docs, and changelog.

## Checklist

- [x] All tests pass (except known PE failure)
- [x] Code compiles cleanly with flags `-Wall -Wpedantic -Wconversion -Wextra`.
- [x] The new code follows GridKit™ style guidelines.
- [x] There are unit tests for the new code.
- [x] The new code is documented.
- [x] The feature branch is rebased with respect to the target branch.
- [x] I have updated [CHANGELOG.md](/CHANGELOG.md) to reflect the changes in this PR.

## Further comments (A)

@nkoukpaizan, I keep running into an issue with Enzyme for Exciter models in the control voltage effective input algebraic equation:
```code
const ScalarT ref_err = vref_ - ev;
const ScalarT sig_err = vs - vc;
const ScalarT lim_err = sUEL_ * vuel - vf;
f[EV]                 = ref_err + sig_err + lim_err;
```

Enzyme fails if that is an inline expression. Sparsity issue. I am not sure if this is intentional, but I finally found a reliable way to circumvent the issue within the model implementation (the code above).

## Further comments (B)

@pelesh @abirchfield @alexander-novo 

Out of PR scope, but there is a modeling issue that I'd like to raise here:
- The 'Compensation voltage', $E_C$, typically comes from the machine model instead of the terminal Bus. This is because there is a parameter specified by the machine model (e.g.`GENROU`) that we do not currently support: $R_{comp}$ and $X_{comp}$. When these are both zero-valued, $E_C=|V_{bus}|$. Sometimes, the compensation reactance is configured to be nonzero specifically to decouple $E_C$ from $|V_{bus}|$ and improve numerical stability. I think there is a broader conversation to be had about how we can model sensed terminal voltage. 
- This exciter as well as  `IEEET1`, locally compute $E_C=\sqrt{V_r^2+V_i^2}$, despite $E_C$ being documented as an external variable.  This was a temporary design choice before we had signals. 
-  For more advanced grid cases and models, the terminal voltage magnitude is consumed more than once. Therefore, I encourage us to introduce $V_{m}$ as a third internal algebraic variable of `Bus`,  conditionally included based on consumption (otherwise we could introduce unused nonlinear algebraic equations incessantly!). Same for $\theta_{bus}$. I will soon need models that need the bus frequency (i.e., time derivative of $\theta_{bus}$).

Please let me know if I should open distinct issues for these. They are front of mind as I implemented this so I am leaving here for now.

### CODE REVIEW

**Comment by lukelowry** [comment]
Rebased and cleaned. Issue discussed in PR body is still an issue but out of the scope of this PR

**Comment by pelesh** [comment]
Needs rebasing.

**Review by pelesh — APPROVED** [review]
Looks good to me. Before we merge, I'd like @nkoukpaizan and @abirchfield to chime in.

@lukelowry: Please create a separate issue for compensation voltage issue you raised in this PR but qualified it as out of scope.

File: GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1a.hpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        }
+
+        const Model::VariableMonitorBase* getMonitor() const override;
+
+        __attribute__((always_inline)) inline int evaluateInternalResidual(
+            const ScalarT*, const ScalarT*, const ScalarT*, const ScalarT*, ScalarT*);
+
+      private:
+        using I = InternalVariablesT;
+        using E = ExternalVariablesT;
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1a.hpp]: `I` and `E` are not expressive enough in later uses, and they will be harder to search. I recommend using the verbose alias directly.

File: GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp [hunk 2]
```diff
@@ -0,0 +1,782 @@
...
+
+        lim_on_ = ZERO<RealT>;
+        if (exclim_)
+        {
+          lim_on_ = ONE<RealT>;
+        }
+
+        // A disabled or inconsistent saturation curve keeps the zero fit so
+        // the coefficients stay finite; verify() reports inconsistent data.
+        const bool saturation_enabled = !(Se1_ == ZERO<RealT> && Se2_ == ZERO<RealT>);
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp]: A few instances of `== ZERO<RealT`.

File: GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp [hunk 3]
```diff
@@ -0,0 +1,782 @@
...
+          }
+          vll0 = vuel0 + inverseRamp(gate_margin0);
+        }
+
+        const ScalarT vf0   = ScalarT{ZERO<RealT>};
+        const ScalarT ev0   = vll0;
+        const ScalarT xll0  = ev0;
+        const ScalarT vref0 = ev0 + vc0 + vf0 - vs0 - uel_on_ * vuel0;
+
+        y[index(I::EFDP)] = efdp0;
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp]: #415 used a different approach to handle the enum indices. I recommend keeping them consistent. If we decide to update the approach, we want to be able to search and replace.

File: tests/UnitTests/PhasorDynamics/ExciterEsdc1aTests.hpp [hunk 4]
```diff
@@ -0,0 +1,1249 @@
...
+
+        bool success = implicit_defaults.initialize(1.2)
+                       && explicit_defaults.initialize(1.2);
+        if (!success)
+        {
+          std::cout << "ESDC1A documented-default comparison failed to initialize\n";
+          return false;
+        }
+
+        success *= (implicit_defaults.evaluate() == 0);
```
Comment (nkoukpaizan) [inline tests/UnitTests/PhasorDynamics/ExciterEsdc1aTests.hpp]: `gcc` warns about these (multiple instances).
```
GridKit/tests/UnitTests/PhasorDynamics/ExciterEsdc1aTests.hpp:936:17: warning: ‘*’ in boolean context, suggest ‘&&’ instead [-Wint-in-bool-context]
  936 |         success *= (implicit_defaults.evaluate() == 0);
```

**Review by nkoukpaizan — COMMENTED** [review]
A few minor comments.

Should this PR add to `CaseFormatTests` and `SystemSingleComponentTests`?

File: GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1a.hpp (hunk 1 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1a.hpp]: Removed and I removed this in the other branches too. it was more convenient to type but I agree with you here

File: GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp (hunk 2 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp]: Thank you. I think all the other PR branches are free of this too now

File: GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp (hunk 3 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp]: I have made these branches the same convention as REGCA now, so it should be good. I will ensure other are too

**Comment by lukelowry** [comment]
Phew, should be good. I standardized the others as best I could again to fix my inconsistency.

On the bright side there's definitely a huge incentive to extract shared code to make the tests more maintainable. I feel confident these models are foolproof w.r.t. edge cases

File: tests/UnitTests/PhasorDynamics/ExciterEsdc1aTests.hpp (hunk 4 above)
Comment (lukelowry) [reply tests/UnitTests/PhasorDynamics/ExciterEsdc1aTests.hpp]: resolved I believe. thank you

File: tests/UnitTests/PhasorDynamics/ExciterEsdc1aTests.hpp [hunk 5]
```diff
@@ -1100,25 +1100,25 @@ namespace GridKit
...
         setAnswerKeyInputs(implicit_defaults);
         setAnswerKeyInputs(explicit_defaults);
         setAnswerKeyState(implicit_defaults.esdc1a);
         setAnswerKeyState(explicit_defaults.esdc1a);
-        success *= (implicit_defaults.evaluate() == 0);
-        success *= (explicit_defaults.evaluate() == 0);
-        success *= vectorUnchanged(implicit_defaults.esdc1a.getResidual(),
+        success &= (implicit_defaults.evaluate() == 0);
+        success &= (explicit_defaults.evaluate() == 0);
+        success &= vectorUnchanged(implicit_defaults.esdc1a.getResidual(),
```
Comment (pelesh) [inline tests/UnitTests/PhasorDynamics/ExciterEsdc1aTests.hpp]: No this cannot be merged.

**Review by pelesh — CHANGES_REQUESTED** [review]
I don't think using `&=` is a good solution here.

File: tests/UnitTests/PhasorDynamics/ExciterEsdc1aTests.hpp (hunk 5 above)
Comment (lukelowry) [reply tests/UnitTests/PhasorDynamics/ExciterEsdc1aTests.hpp]: Addressed in [9380239](https://github.com/ORNL/GridKit/pull/424/commits/93802396d1219817669f663e464b897ba89175e3)

**Comment by lukelowry** [comment]
@pelesh should be ready!

**Comment by lukelowry** [comment]
TestOutcome return type standardized

File: tests/UnitTests/PhasorDynamics/SystemTests.hpp [hunk 6]
```diff
@@ -159,6 +162,53 @@ namespace GridKit
...
 
+      /// GENROU seeds the shared field-voltage node before ESDC1A consumes
+      /// and preserves that value during system initialization.
+      TestOutcome esdc1aInitializationHandoff()
+      {
+        using MachineExternal = PhasorDynamics::GenrouExternalVariables;
+        using ExciterInternal = PhasorDynamics::Exciter::Esdc1aInternalVariables;
+        using ExciterParams   = PhasorDynamics::Exciter::Esdc1aParameters;
+
+        constexpr RealT kTol = static_cast<RealT>(1.0e-9);
```
Comment (nkoukpaizan) [inline tests/UnitTests/PhasorDynamics/SystemTests.hpp]: Please add a comment here.

File: tests/UnitTests/PhasorDynamics/runSystemTests.cpp [hunk 7]
```diff
@@ -10,6 +10,7 @@ int main()
 
   result += test.constructor();
   result += test.composer();
+  result += test.esdc1aInitializationHandoff();
```
Comment (nkoukpaizan) [inline tests/UnitTests/PhasorDynamics/runSystemTests.cpp]: I'm not convinced this is the best spot for this test. Consider something like a new file with `machine/exciter tests` or `twoComponentTests`, where `GastPtiIntegrationTests` in #420 would fit in as well.

File: GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp [hunk 8]
```diff
@@ -0,0 +1,896 @@
...
+        check(Tc_ >= ZERO<RealT>, "Tc must be non-negative");
+        check(Vrmin_ <= Vrmax_, "Vrmin must be less than or equal to Vrmax");
+        check(UEL_ >= static_cast<IdxT>(0) && UEL_ <= static_cast<IdxT>(3),
+              "UEL must be 0, 1, 2, or 3");
+
+        // Saturation is enabled when either coefficient deviates from zero
+        // in any direction; load_real() has already rejected non-finite
+        // values, so the sign tests are exact.
+        auto deviates_from_zero = [](RealT value)
+        { return value < ZERO<RealT> || value > ZERO<RealT>; };
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp]: At this point `== ZERO<RealT` might be clearer than splitting `isfinite`, `>` and `<`. If you're sure an exact comparison to 0 from json-loaded values is what you want to test, I'm okay with merging a version with `== ZERO<RealT`. That will be easier to search and replace later than the currently proposed solution.

**Review by nkoukpaizan — COMMENTED** [review]
This is close to being good to merge. I'll live with the `== ZERO`/`!= ZERO` for now.

I've looked at the other PRs. Let's merge this one with the cleaner implementation first before updating the others.

File: GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp (hunk 8 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp]: Yes, that is the intent. It is unfortunate, but much of the data for these models uses an exact `0.0` for numeric fields to represent a bypassed or disabled feature.

I will correct this and the other models to use `==` so its cleaner, if you are okay with that.

Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp]: Resolved in [eb58015](https://github.com/ORNL/GridKit/pull/424/commits/eb5801520a9ddcc2c09c77cce37410dcdbdaf567)

File: tests/UnitTests/PhasorDynamics/runSystemTests.cpp (hunk 7 above)
Comment (lukelowry) [reply tests/UnitTests/PhasorDynamics/runSystemTests.cpp]: I made a new file for `ComponentConnectionTests` for this flavor of tests

Comment (lukelowry) [reply tests/UnitTests/PhasorDynamics/runSystemTests.cpp]: [eb58015](https://github.com/ORNL/GridKit/pull/424/commits/eb5801520a9ddcc2c09c77cce37410dcdbdaf567)

File: tests/UnitTests/PhasorDynamics/SystemTests.hpp (hunk 6 above)
Comment (lukelowry) [reply tests/UnitTests/PhasorDynamics/SystemTests.hpp]: Adjusted in [9ee9859](https://github.com/ORNL/GridKit/pull/424/commits/9ee9859fd89900f8fa9e452e32609918e301b631), which is now exactly epsilon. It's difficult to decide when the tolerances should be looser or as tight as possible, so I have not been very careful with this. I added a comment and defined kTol exactly as `epsilon()` since this is one where we one it as tight as possible

**Comment by lukelowry** [comment]
The model is ready @abirchfield @nkoukpaizan

**Review by abirchfield — APPROVED** [review]
Reviewed and discussed in person. Ready to merge.

**Comment by lukelowry** [comment]
Rebased!

File: tests/IntegrationTests/PhasorDynamics/CMakeLists.txt [hunk 9]
```diff
@@ -8,4 +8,11 @@ add_test(
   COMMAND test_pd_integration
   WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/examples/PhasorDynamics/Tiny)
 
-install(TARGETS test_pd_integration)
+add_executable(test_pd_component_connection runComponentConnectionTests.cpp)
```
Comment (nkoukpaizan) [inline tests/IntegrationTests/PhasorDynamics/CMakeLists.txt]: I would argue this is still a unit test (everything within `GridKit::PhasorDynamics`), but that's fairly minor.

If kept in `IntegrationTests`, I think you'll need to move the `if(TARGET SUNDIALS::idas)` from the upper directory here to guard `PhasorDynamicsIntegrationTest` but not `PhasorDynamicsComponentConnectionTest`.

File: GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp [hunk 10]
```diff
@@ -0,0 +1,891 @@
...
+      }
+
+      /**
+       * @brief Invert the smooth CommonMath ramp
+       *
+       * Initialization seeds the inactive high-value gate with the gate
+       * *input*, so the residual reproduces the requested output through the
+       * same smooth ramp it evaluates. Beyond the softplus width the smooth
+       * ramp is the identity to double precision, so the output is returned
+       * unchanged there.
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Exciter/ESDC1A/Esdc1aImpl.hpp]:
```suggestion
       * Initialization seeds the inactive high-value gate with the gate
       * *input*, so the residual reproduces the requested output through the
       * same smooth ramp it evaluates. 
       * 
       * For large positive values, the ramp is effectively equal to the input, so the
       * inverse is effectively the output. In that regime this function returns `ramp_output` directly.
       * This branching is numerically more robust.
```

**Review by nkoukpaizan — APPROVED** [review]
Good enough to merge. Thanks for the hard work!

File: tests/IntegrationTests/PhasorDynamics/CMakeLists.txt (hunk 9 above)
Comment (lukelowry) [reply tests/IntegrationTests/PhasorDynamics/CMakeLists.txt]: I'll move it back to UnitTests; I don't have enough experience to know what a good separation is.

give me one moment I'm doing it right now

Comment (lukelowry) [reply tests/IntegrationTests/PhasorDynamics/CMakeLists.txt]: Moved in [8681821](https://github.com/ORNL/GridKit/pull/424/commits/86818218f220ebad9cd0196dd6fc4942a1fd43b8)

**Comment by lukelowry** [comment]
Good to go. I'll begin making the other open PR models align with this and REGCA. Not opposed if you want to merge #521 after this

**Review by pelesh — APPROVED** [review]
My prior request is addressed.
